### PR TITLE
Feature/stream settings

### PR DESCRIPTION
**Pull request by mja00**
Added stream settings to the settings area.
Allows users to:
- Create a channel(can be used later to review channels before creation)
- Set stream title and category(updates in realtime like the settings on the stream page itself)
- Set text for the chat rules section on the stream page(doesn't update in realtime but it shouldn't be updated often)
- Delete channel(soft deletion and only makes the channel inaccessible to the get_channel_from_username function)

### CODE REVIEW

**Comment by mja00** [comment]
I still have to write tests for all of this but it is currently functional

File: lib/glimesh/streams.ex [hunk 1]
```diff
@@ -94,14 +95,22 @@ defmodule Glimesh.Streams do
     Repo.get_by(Channel, user_id: user.id) |> Repo.preload([:category, :user])
   end
 
-  def create_channel(user, attrs \\ %{}) do
+  def create_channel(user, attrs \\ %{category_id: 1}) do
```
Comment (clone1018) [inline lib/glimesh/streams.ex:98]: What is category id 1?

File: lib/glimesh/streams/channel.ex [hunk 2]
```diff
@@ -28,4 +40,18 @@ defmodule Glimesh.Streams.Channel do
...
     end
   end
+
+  def set_chat_rules_content_html(changeset) do
+    case changeset do
+      %Ecto.Changeset{valid?: true, changes: %{chat_rules_md: chat_rules_md}} ->
+        put_change(
+          changeset,
+          :chat_rules_html,
+          Glimesh.Accounts.Profile.safe_user_markdown_to_html(chat_rules_md)
```
Comment (clone1018) [inline lib/glimesh/streams/channel.ex:50]: Who wrote this cool function?

File: lib/glimesh_web/controllers/user_settings_controller.ex [hunk 3]
```diff
@@ -35,11 +37,68 @@ defmodule GlimeshWeb.UserSettingsController do
...
   end
 
+  def create_channel(conn, _params) do
+    user = conn.assigns.current_user
+
+    case Streams.create_channel(user) do
+      {:ok, channel} ->
+        conn
+        |> put_session(:user_return_to, Routes.user_settings_path(conn, :stream))
+        |> UserAuth.log_in_user(user)
```
Comment (clone1018) [inline lib/glimesh_web/controllers/user_settings_controller.ex:47]: Huh? should this be here?

File: lib/glimesh_web/templates/user_settings/stream.html.eex [hunk 4]
```diff
@@ -1 +1,52 @@
-<h2>If you could stream, your settings would go here.</h2>
+<div class="container">
+    <h2 class="mt-4"><%= gettext("Stream Settings") %></h2>
+    <%= if !@channel do %>
+        <h5><%= gettext("You currently don't have a channel setup. Press the button below if you'd like to setup your channel!") %></h5>
+        <%= button gettext("Create Channel"), class: "btn btn-primary", to: Routes.user_settings_path(@conn, :create_channel), method: "put"%>
```
Comment (clone1018) [inline lib/glimesh_web/templates/user_settings/stream.html.eex:5]: We can do it in another PR, but we might want to just consider making this the create channel page if they don't have one already.

File: lib/glimesh_web/templates/user_settings/stream.html.eex [hunk 5]
```diff
@@ -1 +1,52 @@
...
+            </div>
+            <div class="row">
+                <div class="col-sm-6">
+                    <%= submit gettext("Update Settings"), class: "btn btn-primary" %>
+                </div>
+            </div>
+        </div>
+    </div>
+    <% end %>
+    <%= button("Delete Channel", to: Routes.user_settings_path(@conn, :delete_channel), method: "put", class: "btn btn-danger mt-4", data: [confirm: "Pressing OK will: \n - Delete your channel \n - Wipe all settings tied to the channel \n - Clear all followers for the channel \n\nPressing OK will NOT: \n- Delete your Glimesh account \n- Remove all profile related settings \n\nAre you sure you want to do this?"]) %>
```
Comment (clone1018) [inline lib/glimesh_web/templates/user_settings/stream.html.eex:50]: Should this confirm be translated?

File: test/glimesh/streams_test.exs [hunk 6]
```diff
@@ -125,7 +125,7 @@ defmodule Glimesh.StreamsTest do
 
     test "list_categories/0 returns all categories" do
       category = category_fixture()
-      assert Enum.map(Streams.list_categories(), fn x -> x.name end) == [category.name]
+      assert Enum.member?(Enum.map(Streams.list_categories(), fn x -> x.name end), category.name)
```
Comment (clone1018) [inline test/glimesh/streams_test.exs:128]: neat

File: test/glimesh_web/controllers/user_registration_controller_test.exs [hunk 7]
```diff
@@ -40,7 +40,7 @@ defmodule GlimeshWeb.UserRegistrationControllerTest do
       conn = get(conn, "/")
       response = html_response(conn, 200)
       assert response =~ username
-      assert response =~ "\nSettings</a>"
+      assert response =~ "\nYour Settings</a>"
```
Comment (clone1018) [inline test/glimesh_web/controllers/user_registration_controller_test.exs:43]: Trying to get away from "Your x", can we change this back?

File: test/glimesh_web/controllers/user_session_controller_test.exs [hunk 8]
```diff
@@ -36,7 +36,7 @@ defmodule GlimeshWeb.UserSessionControllerTest do
       conn = get(conn, "/")
       response = html_response(conn, 200)
       assert response =~ user.username
-      assert response =~ "\nSettings</a>"
+      assert response =~ "\nYour Settings</a>"
```
Comment (clone1018) [inline test/glimesh_web/controllers/user_session_controller_test.exs:39]: Same as above, thanks!

File: lib/glimesh/streams/channel.ex [hunk 9]
```diff
@@ -12,13 +12,25 @@ defmodule Glimesh.Streams.Channel do
...
+    |> cast(attrs, [
+      :title,
+      :category_id,
+      :language,
+      :thumbnail,
+      :stream_key,
+      :chat_rules_md,
+      :inaccessible
+    ])
+    |> set_chat_rules_content_html()
```
Comment (clone1018) [inline lib/glimesh/streams/channel.ex:33]: Can we add a validator for `chat_rules_md` to ensure it doesn't get too long?

**Comment by clone1018** [comment]
Awesome stuff man! This puts us in a great place, just left a couple of minor comments.

File: lib/glimesh/streams.ex (hunk 1 above)
Comment (mja00) [reply lib/glimesh/streams.ex:98]: It's currently the Art category. I had to populate it due to when you created your channel with the "Create Channel" button your stream page would break if you didn't set anything.

File: lib/glimesh_web/controllers/user_settings_controller.ex (hunk 3 above)
Comment (mja00) [reply lib/glimesh_web/controllers/user_settings_controller.ex:47]: Currently use it to "reload" the page with the updated information. I'm probably missing a better function to do that.

File: lib/glimesh_web/templates/user_settings/stream.html.eex (hunk 5 above)
Comment (mja00) [reply lib/glimesh_web/templates/user_settings/stream.html.eex:50]: I knew I forgot a gettext somewhere

File: test/glimesh_web/controllers/user_registration_controller_test.exs (hunk 7 above)
Comment (mja00) [reply test/glimesh_web/controllers/user_registration_controller_test.exs:43]: I just changed the test since someone else change it backed to Your Settings in the navbar.

File: lib/glimesh/streams/channel.ex (hunk 9 above)
Comment (mja00) [reply lib/glimesh/streams/channel.ex:33]: I'll need some help on this because currently if a changeset is invalid it breaks the entire page due to the Repo.preload in the update_channel function.

File: test/glimesh_web/controllers/user_registration_controller_test.exs (hunk 7 above)
Comment (mja00) [reply test/glimesh_web/controllers/user_registration_controller_test.exs:43]: It's most likely due to me importing the English translation file into the project. I'll take a look though.

File: lib/glimesh_web/templates/user_settings/stream.html.eex (hunk 4 above)
Comment (mja00) [reply lib/glimesh_web/templates/user_settings/stream.html.eex:5]: Let's do it in a different PR. That way it doesn't delay this one any more than it already is since I would wanna spend some decent time on making that page pretty. :)

File: lib/glimesh/streams.ex (hunk 1 above)
Comment (mja00) [reply lib/glimesh/streams.ex:98]: Changed this to get the list of categories and just pick the first one it finds. It's needed to be set on channel creation otherwise the page will break since category_id would be null.

File: lib/glimesh/streams/channel.ex (hunk 9 above)
Comment (mja00) [reply lib/glimesh/streams/channel.ex:33]: Added back the validate clause but it will throw an internal error if the md is over 8192 characters. Will have to figure a way of showing the error generated.

Comment (mja00) [reply lib/glimesh/streams/channel.ex:33]: No longer errors, did some error checking internally since the function is shared between two pages.
